design/logging.md

```diff
@@ -0,0 +1,495 @@
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/callbacks.rb:28:in `block in call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/callbacks.rb:98:in `run_callbacks'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/debug_exceptions.rb:61:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:38:in `call_app'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:26:in `block in call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:71:in `block in tagged'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:28:in `tagged'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/tagged_logging.rb:71:in `tagged'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/rack/logger.rb:26:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/request_id.rb:27:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/method_override.rb:24:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/runtime.rb:22:in `call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/static.rb:127:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/sendfile.rb:110:in `call'
+/var/lib/gems/2.5.0/gems/conjur-rack-heartbeat-2.2.0/lib/rack/heartbeat.rb:20:in `call'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/engine.rb:524:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/configuration.rb:227:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:675:in `handle_request'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:476:in `process_client'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:334:in `block in run'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
+ INFO 2020/05/17 17:38:07 +0000 [pid=300] [origin=127.0.0.1] [request_id=d3f54689-ed9d-4c66-858a-d01d9c7adbc9] [tid=317] Completed 401 Unauthorized in 447ms
+```
+
+Here we can see that the log is more robust and has all the data that can help a developer
+to investigate the issue, including the stack trace and the DB messages.
+
+### Backwards compatibility
+
+It's important to say that the UX will not change by default. We will do this by changing
+the default value for the log level from `info` to `warn`. This way the log will show the
+same data that it shows today.
```

Comment:
Reading this again now and I see that this is incorrect. The UX will be different if we change the default log level to `warn`. For example, in the use-case described above with `authn-oidc` the log will be empty and will not include these log messages:
```
 INFO 2020/05/18 09:49:20 +0000 [pid=411] [origin=127.0.0.1] [request_id=0487ea6a-a974-481a-8b7f-d7d62e96aec6] [tid=421] Started POST "/authn-oidc/keycloak/cucumber/authenticate" for 127.0.0.1 at 2020-05-18 09:49:20 +0000
 INFO 2020/05/18 09:49:21 +0000 [pid=411] [origin=127.0.0.1] [request_id=0487ea6a-a974-481a-8b7f-d7d62e96aec6] [tid=421] Processing by AuthenticateController#authenticate_oidc as */*
 INFO 2020/05/18 09:49:21 +0000 [pid=411] [origin=127.0.0.1] [request_id=0487ea6a-a974-481a-8b7f-d7d62e96aec6] [tid=421]   Parameters: {"service_id"=>"keycloak", "account"=>"cucumber"}
 INFO 2020/05/18 09:49:21 +0000 [pid=411] [origin=127.0.0.1] [request_id=0487ea6a-a974-481a-8b7f-d7d62e96aec6] [tid=421] Completed 401 Unauthorized in 94ms
```
However, before this change it was printed by default without changing the log level. 

We have 2 options here:
1. Still change the default level to `warn` and change the current log behaviour
2. Leave the default log level at `info` which will add more logs to the current log as it will include the logs that we currently log in DEBUG (all the `CONJ` messages).

Option 2 is better in my opinion and will have the following implications:
1. We will not need to make any changes in `evoke`, `conjur-ui` & `ldap-sync`
2. We will need to get a security approval for this change as more data is written to the logs
3. The UX will change as the log will have more log messages in it by default. However, we don't have many logs added as we will add only our logs and not the DB and stacktrace.

@rafis3 wdyt?

Reply:
@orenbm I like the second option better as well. My main concern is that more logs means more space required on disk. For a loaded environment, this will produce more output that the Conjur/DAP instances will need to have enough space to handle. It also means more I/O which can affect performance on very high loads. So in a production environment, I would recommend keeping it at `warn` unless there is a problem. But I'd prefer that by default, when you start working with the product, you'd get an out of the box log level that actually helps you understand the product and solve problems.